File: back-end/src/infrastructure/database/models/event_model.py
from tortoise.models import Model
from tortoise import fields
from ....domain.entities.event import EventStatus


class EventModel(Model):
    """Tortoise ORM model for Event entity"""
    id = fields.IntField(pk=True)
    title = fields.CharField(max_length=200)
    description = fields.TextField()
    venue = fields.CharField(max_length=200)
    date_time = fields.DatetimeField()
    capacity = fields.IntField()
    price = fields.DecimalField(max_digits=10, decimal_places=2)
    status = fields.CharEnumField(EventStatus, default=EventStatus.ACTIVE)
    created_at = fields.DatetimeField(auto_now_add=True)
    
    # Computed fields for performance (updated by triggers)
    total_tickets_sold = fields.IntField(default=0)
    total_revenue = fields.DecimalField(max_digits=10, decimal_places=2, default=0)
    total_bookings = fields.IntField(default=0)
    
    class Meta:
        table = "events"
    
    def __str__(self):
        return f"{self.title} - {self.venue}"

    @property
    def available_tickets(self) -> int:
        """Calculate available tickets"""
        return self.capacity - self.total_tickets_sold
    
    @property
    def occupancy_percentage(self) -> float:
        """Calculate occupancy percentage"""
        if self.capacity == 0:
            return 0.0
        return round((self.total_tickets_sold * 100.0) / self.capacity, 2)
    
    @property
    def potential_revenue(self) -> float:
        """Calculate potential revenue if all tickets are sold"""
        return float(self.capacity * self.price)

File: front-end/src/app/core/models/event.model.ts
export interface Event {
    id: number
    title: string
    description: string
    venue: string
    date: Date
    time: string
    price: number
    totalTickets: number
    availableTickets: number
    image: string
    isActive: boolean
    createdAt: Date
}

export interface EventManagement {
    id: number
    title: string
    description: string
    venue: string
    date_time: string
    capacity: number
    price: string
    status: string
    created_at: string
    total_tickets_sold: number
    available_tickets: number
    total_revenue: string
    total_bookings: number
    occupancy_percentage: number
    potential_revenue: string
}

export interface EventFilters {
    search?: string
    venue?: string
    dateFrom?: Date
    dateTo?: Date
}
